File: 2023/day2/Day2.hs
{-# LANGUAGE OverloadedStrings #-}

import Data.Char (isDigit, isSpace)
import Data.List (foldl')
import Data.Monoid ((<>))
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Text.IO as TIO
import Data.Text.Lazy (toStrict)
import Data.Text.Lazy.Builder (Builder, toLazyText)
import Data.Text.Lazy.Builder.Int (decimal)
import qualified Data.Text.Read as TR (decimal)

data Color = Red | Green | Blue

data Game = Game {no :: Int, red :: [Int], green :: [Int], blue :: [Int]}

main :: IO ()
main = TIO.interact solution

solution :: Text -> Text
solution t = toStrict $ toLazyText ("Part1: " <> p1 <> "\n" <> "Part2: " <> p2 <> "\n" :: Builder)
  where
    ls = T.lines t
    games = map parseGame ls
    p1 = decimal $ part1 games
    p2 = decimal $ part2 games

part1 :: [Game] -> Int
part1 = sum . map no . filter possible

part2 :: [Game] -> Int
part2 = sum . map gamePower

gamePower :: Game -> Int
gamePower (Game _ rs gs bs) = product [maximum rs, maximum gs, maximum bs]

parseGame :: Text -> Game
parseGame t = case TR.decimal $ T.drop 5 t of
    Left _ -> error "error parsing game"
    Right (n, rest) -> case parseHands $ T.drop 2 rest of
        (rs, gs, bs) -> Game n rs gs bs

parseHands :: Text -> ([Int], [Int], [Int])
parseHands = foldl' accFun ([], [], []) . concatMap (hands . T.split (',' ==)) . T.split (';' ==)
  where
    accFun (rs, gs, bs) (n, c) = case c of
        Red -> (n : rs, gs, bs)
        Green -> (rs, n : gs, bs)
        Blue -> (rs, gs, n : bs)
        
    hands = map (\t -> (cnt t, color t))

    cnt :: Text -> Int
    cnt = either (error "no count") fst . TR.decimal . T.takeWhile isDigit . T.dropWhile isSpace

    color t = case T.last $ T.stripEnd t of
        'd' -> Red
        'n' -> Green
        'e' -> Blue
        _ -> error ":O"

possible :: Game -> Bool
possible (Game _ rs gs bs) =
    let
        maxR = maximum rs
        maxG = maximum gs
        maxB = maximum bs
     in
        maxR <= 12 && maxG <= 13 && maxB <= 14

File: 2023/day2/Day2.scala
import scala.io.StdIn.readLine

enum Color:
  case Red, Green, Blue

case class Game(id: Int, hands: Map[Color, Array[Int]])
case class GameMax(id: Int, red: Int, green: Int, blue: Int)

@main def main =
    val gameMaxes = 
        Iterator.continually(readLine())
            .takeWhile(s => s != null && !s.isEmpty())
            .map(parseGame)
            .map(gameMax)
            .toArray

    val part1 = gameMaxes.filter(possible).map(_.id).sum
    val part2 = gameMaxes.map(g => g.red * g.green * g.blue).sum

    println(s"Part1: $part1\nPart2: $part2")

def parseGame(s: String): Game =    
    def number(span: String) = span.dropWhile(!_.isDigit).toInt
    
    def hand(s: String): Array[(Int, Color)] =        
        s.split(',').map: x =>
            val (cnt, rest) = x.strip().span(!_.isSpaceChar)
            val color =
                rest.drop(1) match
                    case "red"   => Color.Red
                    case "green" => Color.Green
                    case "blue"  => Color.Blue
                    case err     => throw new RuntimeException(s"Impossible color: $err")
            (cnt.toInt, color)
            
    def hands(s: String): Map[Color, Array[Int]] =
        s.drop(1).split(';').map(hand).flatten.groupMap(_._2)(_._1)
    
    s.span(_ != ':') match
        case (game, rest) => Game(number(game), hands(rest))
    
def gameMax(game: Game): GameMax =
    GameMax(game.id, 
        game.hands(Color.Red).max, 
        game.hands(Color.Green).max, 
        game.hands(Color.Blue).max)

def possible(g: GameMax): Boolean =
       g.red   <= 12
    && g.green <= 13 
    && g.blue  <= 14
